stl/inc/span

```diff
@@ -0,0 +1,685 @@
+    }
+
+    constexpr _Span_iterator operator++(int) noexcept {
+        _Span_iterator _Tmp{*this};
+        ++*this;
+        return _Tmp;
+    }
+
+    constexpr _Span_iterator& operator--() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(_Mybegin, "cannot decrement value-initialized span iterator");
+        _STL_VERIFY(_Mybegin < _Myptr, "cannot decrement span iterator before begin");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        --_Myptr;
+        return *this;
+    }
+
+    constexpr _Span_iterator operator--(int) noexcept {
+        _Span_iterator _Tmp{*this};
+        --*this;
+        return _Tmp;
+    }
+
+    constexpr void _Verify_offset([[maybe_unused]] const difference_type _Off) const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        if (_Off != 0) {
+            _STL_VERIFY(_Mybegin, "cannot seek value-initialized span iterator");
+        }
+
+        if (_Off < 0) {
+            _STL_VERIFY(_Myptr - _Mybegin >= -_Off, "cannot seek span iterator before begin");
+        }
+
+        if (_Off > 0) {
+            _STL_VERIFY(_Myend - _Myptr >= _Off, "cannot seek span iterator after end");
+        }
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+    }
+
+    constexpr _Span_iterator& operator+=(const difference_type _Off) noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _Verify_offset(_Off);
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        _Myptr += _Off;
+        return *this;
+    }
+
+    _NODISCARD constexpr _Span_iterator operator+(const difference_type _Off) const noexcept {
+        _Span_iterator _Tmp{*this};
+        _Tmp += _Off;
+        return _Tmp;
+    }
+
+    constexpr _Span_iterator& operator-=(const difference_type _Off) noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        if (_Off != 0) {
+            _STL_VERIFY(_Mybegin, "cannot seek value-initialized span iterator");
+        }
+
+        if (_Off > 0) {
+            _STL_VERIFY(_Myptr - _Mybegin >= _Off, "cannot seek span iterator before begin");
+        }
+
+        if (_Off < 0) {
+            _STL_VERIFY(_Myend - _Myptr >= -_Off, "cannot seek span iterator after end");
+        }
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        _Myptr -= _Off;
+        return *this;
+    }
+
+    _NODISCARD constexpr _Span_iterator operator-(const difference_type _Off) const noexcept {
+        _Span_iterator _Tmp{*this};
+        _Tmp -= _Off;
+        return _Tmp;
+    }
+
+    _NODISCARD constexpr difference_type operator-(const _Span_iterator& _Right) const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(
+            _Mybegin == _Right._Mybegin && _Myend == _Right._Myend, "cannot subtract incompatible span iterators");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        return _Myptr - _Right._Myptr;
+    }
+
+    _NODISCARD constexpr reference operator[](const difference_type _Off) const noexcept {
+        return *(*this + _Off);
+    }
+
+    _NODISCARD constexpr bool operator==(const _Span_iterator& _Right) const noexcept {

```

Comment:
It looks like comparing `const_iterator == iterator` will compile (converting the RHS to `const_iterator`), but comparing `iterator == const_iterator` will not compile. Did you test that scenario?

Reply:
I recommend performing this test with Clang, because I'm relatively certain compilers like MSVC that implement WG21-P1185 will make this work via magic rewrites. (We also need to support Clang 9 and EDG 5.1 which do not.)

Reply:
In the [unit tests](https://github.com/miscco/SpanTest/blob/7b338e6d494e553ff1bd6d170e8d337cbb394200/SpanTest/begin.cpp#L28) for `begin()`, `end()`, `rbegin()` and `rend()` there is always the condition `iterator == const_iterator`

However, it might work due to magic